### PR TITLE
feat: Generate Ignore.xml

### PR DESCRIPTION
**Pull request by workgroupengineering**
GitHub Issue (If applicable): #

## PR Type
What kind of change does this PR introduce?

- Feature

## What is the new behavior?

Allowed to generate Ingnore.xml

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal)

## Other information

Internal Issue (If applicable):

Part of https://github.com/unoplatform/uno/issues/10885

### CODE REVIEW

File: src/Uno.PackageDiff/GitHubClient.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
+﻿using System;
+using System.Text;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using System.Web;
+
+namespace Uno.PackageDiff
+{
+	internal class GitHubClient
```
Comment (Xiaoy312) [inline src/Uno.PackageDiff/GitHubClient.cs:8]: might be overkill, but we could just plug in an existing api library like `Octokit`?

File: src/Uno.PackageDiff/GitHubClient.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+		private static async Task<dynamic> PostDocument(Uri contentUri, string githubPAT, string document)
+		{
+			var wc = new System.Net.WebClient();
+			wc.Headers.Add("User-agent", "uno-nv-sync");
+			wc.Headers.Add("Authorization", $"token {githubPAT}");
+			wc.Encoding = UTF8Encoding.UTF8;
+
+			var result = await wc.UploadStringTaskAsync(contentUri, "POST", document);
+
+			return JsonConvert.DeserializeObject(result);
```
Comment (Xiaoy312) [inline src/Uno.PackageDiff/GitHubClient.cs]: `dynamic` seem like a worse idea, than just returning the `string`
but, if you arent using the result pass debugging stage
it is better to just let this be a `Task` _of nothing_

File: src/Uno.PackageDiff/GitHubClient.cs (hunk 1 above)
Comment (workgroupengineering) [reply src/Uno.PackageDiff/GitHubClient.cs:8]: I had thought about it, but it seemed excessive to add Octokit just to create a PR comment.

File: src/Uno.PackageDiff/IDiffWriter.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
+﻿using System;
+using NuGet.Versioning;
+using Mono.Cecil;
+
+namespace Uno.PackageDiff
+{
+	internal interface IDiffWriter : IDisposable
```
Comment (Xiaoy312) [inline src/Uno.PackageDiff/IDiffWriter.cs:7]: this contract looks highly tied to xml syntax (`WriteXyzStart`, `WriteXyzEnd`), and md formatting (XyzStart that takes in an `int`).
it should instead accept the array of missing types, and (containing types + their missing members: properties,fields,method and events)
or anything more context-relevant

File: src/Uno.PackageDiff/Program.cs [hunk 4]
```diff
@@ -18,26 +18,77 @@
...
+				//
+				{ "github-pat=", s => githubPAT = s },
+				{ "source-repository=", s => sourceRepository = s },
+				{ "github-pr-id=", s => githubPRid = s  }
 			};
 
 			p.Parse(args);
 
+			IDiffWriter writer = null;
+			if(string.IsNullOrEmpty(outputType))
```
Comment (Xiaoy312) [inline src/Uno.PackageDiff/Program.cs:54]: the if-else can be moved to a separate method:
```cs
using (var writer = BuildWriter(outtype, /* github params... */))
```

File: src/Uno.PackageDiff/Writers/GitHubWriter.cs [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+		{
+			using var ms = new System.IO.MemoryStream();
+			using var sw = new System.IO.StreamWriter(ms);
+
+			sw.WriteLine($"# Comparison report for {source.Package} **{source.Version}** with {target.Package} **{target.Version}**");
+			sw.WriteLine();
+
+			if(types.Count + properties.Count + events.Count + methods.Count + fields.Count > 0)
+			{
+				sw.WriteLine("🚨🚨**detected some breaking changes** if changes are wanted, add following line at end of `PackageDiffIgnore.xml`.🚨🚨");
```
Comment (Xiaoy312) [inline src/Uno.PackageDiff/Writers/GitHubWriter.cs:98]: the sections can be long
you might want to wrap them in \<details> syntax:
<details>
<summary>like so</summary>

~~~xml
<details>
<summary>like so</summary>

```xml
...// note: you need to leave a empty line between <summary> and start of ```
```
</details>
~~~
</details>

File: src/Uno.PackageDiff/Writers/XmlIngnoreWriter.cs [hunk 6]
```diff
@@ -0,0 +1,165 @@
+﻿using System.Collections.Generic;
+using System.IO;
+using System.Xml;
+using Mono.Cecil;
+using NuGet.Versioning;
+
+namespace Uno.PackageDiff.Writers
+{
+	internal class XmlIngnoreWriter : IDiffWriter
```
Comment (Xiaoy312) [inline src/Uno.PackageDiff/Writers/XmlIngnoreWriter.cs:9]:
```suggestion
	internal class XmlIgnoreWriter : IDiffWriter
```

Comment (Xiaoy312) [reply src/Uno.PackageDiff/Writers/XmlIngnoreWriter.cs:9]: same for commit+PR title too

File: src/Uno.PackageDiff/IDiffWriter.cs (hunk 3 above)
Comment (Xiaoy312) [inline src/Uno.PackageDiff/IDiffWriter.cs:7]: this contract looks highly tied to xml syntax (`WriteXyzStart`, `WriteXyzEnd`), and md formatting (XyzStart that takes in an `int`).
it should instead accept the array of missing types, and (containing types + their missing members: properties,fields,method and events)
or anything more context-relevant

File: src/Uno.PackageDiff/Program.cs (hunk 4 above)
Comment (Xiaoy312) [inline src/Uno.PackageDiff/Program.cs:54]: the if-else can be extracted to a separate method:
```cs
using (var writer = BuildWriter(outtype, /* github params... */))
```

File: src/Uno.PackageDiff/Program.cs [hunk 7]
```diff
@@ -18,26 +18,77 @@
...
+						else if(string.Equals("github", type, StringComparison.OrdinalIgnoreCase))
+						{
+							composite.Add(new Writers.GitHubWriter(githubPAT, sourceRepository, githubPRid));
+						}
+					}
+					if(composite.Count == 0)
+					{
+						Console.WriteLine($"Error : Invalid outtype {outputType}.");
+						return 1;
+					}
```
Comment (Xiaoy312) [inline src/Uno.PackageDiff/Program.cs:87]: a combination of valid and invalid `outtype` will slip through here... ex: "outtype=diff,asd" 

you should add a final else: `else { Console.WriteLine($"Error: Invalid outtype: {type}"); }`
and change the final if to `composite.Count != types.Length`

File: src/Uno.PackageDiff/Writers/XmlIngnoreWriter.cs (hunk 6 above)
Comment (Xiaoy312) [inline src/Uno.PackageDiff/Writers/XmlIngnoreWriter.cs:9]:
```suggestion
	internal class XmlIgnoreWriter : IDiffWriter
```
and, the same in commit+pr title too

**Comment by workgroupengineering** [comment]
Replaced by #20

**Pull request closed without merging**